odoo/addons/base/res/res_currency.py

```diff
@@ -169,34 +174,38 @@ def is_zero(self, amount):
         return tools.float_is_zero(amount, precision_rounding=self.rounding)
 
     @api.model
-    def _get_conversion_rate(self, from_currency, to_currency):
+    def _get_conversion_rate(self, from_currency, to_currency, company_id, date):
         from_currency = from_currency.with_env(self.env)
         to_currency = to_currency.with_env(self.env)
-        return to_currency.rate / from_currency.rate
+        currency_rates = (from_currency + to_currency).get_rates(company_id, date)
+        return currency_rates.get(to_currency.id) / currency_rates.get(from_currency.id)
 
-    @api.model
-    def _compute(self, from_currency, to_currency, from_amount, round=True):
-        if (to_currency == from_currency):
-            amount = to_currency.round(from_amount) if round else from_amount
-        else:
-            rate = self._get_conversion_rate(from_currency, to_currency)
-            amount = to_currency.round(from_amount * rate) if round else from_amount * rate
-        return amount
-
-    @api.multi
-    def compute(self, from_amount, to_currency, round=True):
-        """ Convert `from_amount` from currency `self` to `to_currency`. """
+    def convert_amount(self, from_amount, to_currency, company_id, date, round=True):
```

Comment:
:+1: on the new API; :-1: on screwing up documentation

Suggestions:
 - name it `convert`; I prefer shorter method names except when they are ambiguous
 - parameter `company_id` should be `company`; forget about record ids, it's old school ;-)
 - I second @xmo-odoo on using keyword-only parameters here as well

The calling idiom should be clear enough:
```python
    result = currency1.convert(amount, currency2, company=my_company, date=some_date)
```